ledger-api/grpc-definitions/com/daml/ledger/api/v2/commands.proto

```diff
@@ -0,0 +1,304 @@
+  // Note: The ledger time of a transaction is assigned as part of command interpretation.
+  // Use this property if you expect that command interpretation will take a considerate amount of time, such that by
+  // the time the resulting transaction is sequenced, its assigned ledger time is not valid anymore.
+  // Must not be set at the same time as min_ledger_time_rel.
+  // Optional
+  google.protobuf.Timestamp min_ledger_time_abs = 8;
+
+  // Same as min_ledger_time_abs, but specified as a duration, starting from the time the command is received by the server.
+  // Must not be set at the same time as min_ledger_time_abs.
+  // Optional
+  google.protobuf.Duration min_ledger_time_rel = 9;
+
+  // Set of parties on whose behalf the command should be executed.
+  // If ledger API authorization is enabled, then the authorization metadata must authorize the sender of the request
+  // to act on behalf of each of the given parties.
+  // This field supersedes the ``party`` field. The effective set of parties on whose behalf the command
+  // should be executed is the union of all parties listed in ``party`` and ``act_as``, which must be non-empty.
+  // Each element must be a valid PartyIdString (as described in ``value.proto``).
+  // Optional
+  repeated string act_as = 10;
+
+  // Set of parties on whose behalf (in addition to all parties listed in ``act_as``) contracts can be retrieved.
+  // This affects Daml operations such as ``fetch``, ``fetchByKey``, ``lookupByKey``, ``exercise``, and ``exerciseByKey``.
+  // Note: A participant node of a Daml network can host multiple parties. Each contract present on the participant
+  // node is only visible to a subset of these parties. A command can only use contracts that are visible to at least
+  // one of the parties in ``act_as`` or ``read_as``. This visibility check is independent from the Daml authorization
+  // rules for fetch operations.
+  // If ledger API authorization is enabled, then the authorization metadata must authorize the sender of the request
+  // to read contract data on behalf of each of the given parties.
+  // Optional
+  repeated string read_as = 11;
+
+  // A unique identifier to distinguish completions for different submissions with the same change ID.
+  // Typically a random UUID. Applications are expected to use a different UUID for each retry of a submission
+  // with the same change ID.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  //
+  // If omitted, the participant or the committer may set a value of their choice.
+  // Optional
+  string submission_id = 12;
+
+  // Additional contracts used to resolve contract & contract key lookups.
+  // Optional
+  repeated DisclosedContract disclosed_contracts = 13;
+
+  // Must be a valid domain ID
+  // Required
+  string domain_id = 14;
+}
+
+// A command can either create a new contract or exercise a choice on an existing contract.
+message Command {
```

Comment:
Duplicating those command types is going to be fairly annoying to work with. That's what the Java codegen prdouces for command submissions (only `Command` not `Commands`). Afaict you're not actually changing any of them. Could we reuse them? Alternatively we'll have to either patch that part of the Java codegen or we hack around it in CN and write our own translation.

I think more generally what I'm missing atm is a good rule for what you duplicate and what you reuse. E.g., what's the reasoning for why CreatedEvent gets reused but not Commands?

Reply:
Yes, I was suspicious of the Command too.
Answering your question I was following a crude approach: whenever I changed a file I moved everything over. Nothing prevents us to move only the changed messages. I will apply that.
The change in the commands file is focusing on the Commands indeed.
So you are saying it would be enough to keep the Command on v1 (...to avoid trouble with Java codegen)?

Reply:
I'd suggest at least until June we keep it yeah. Reasonably confident that's enough but somewhat difficult to say with complete certainty since I obviously cannot just run our code against that API to test it.

For 3.0 adjusting the Java codegen may be a reasonable option if it helps but if it's literally just a 1:1 copy just referencing the old definition seems possibly nicer.

Reply:
I removed the duplication of Command, this is what you had in mind?

Reply:
yes thx